drivers/i2s/i2s_silabs_siwx91x.c

```diff
@@ -0,0 +1,936 @@
+				 (uint32_t)stream->mem_block, stream->cfg.block_size);
+	} else {
+		ret = i2s_siwx91x_prepare_dma_channel(i2s_dev, stream->mem_block,
+						      stream->cfg.block_size, stream->dma_channel,
+						      false);
+		stream->reload_en = false;
+	}
+
+	if (ret < 0) {
+		goto rx_disable;
+	}
+
+	ret = dma_start(dma_dev, stream->dma_channel);
+	if (ret < 0) {
+		goto rx_disable;
+	}
+
+	return;
+
+rx_disable:
+	i2s_siwx91x_stream_disable(stream, dma_dev);
+}
+
+static void dma_tx_callback(const struct device *dma_dev, void *user_data, uint32_t channel,
+			    int status)
+{
+	const struct device *i2s_dev = (const struct device *)user_data;
+	const struct i2s_siwx91x_config *cfg = i2s_dev->config;
+	struct i2s_siwx91x_data *data = i2s_dev->data;
+	struct i2s_siwx91x_stream *stream = &data->tx;
+	size_t mem_block_size;
+	uint8_t data_size; /* data size in bytes */
+	int ret;
+
+	__ASSERT_NO_MSG(stream->mem_block != NULL);
+
+	k_mem_slab_free(stream->cfg.mem_slab, stream->mem_block);
+	stream->mem_block = NULL;
+
+	if (stream->state == I2S_STATE_ERROR) {
+		goto tx_disable;
+	}
+
+	if (stream->last_block) {
+		stream->state = I2S_STATE_READY;
+		goto tx_disable;
+	}
+
+	ret = i2s_siwx91x_queue_get(&stream->mem_block_queue, &stream->mem_block, &mem_block_size);
+	if (ret < 0) {
+		if (stream->state == I2S_STATE_STOPPING) {
+			stream->state = I2S_STATE_READY;
+		} else {
+			stream->state = I2S_STATE_ERROR;
+		}
+		goto tx_disable;
+	}
+
+	k_sem_give(&stream->sem);
+
+	if (stream->cfg.word_size == 24) {
+		data_size = 4;
+	} else {
+		data_size = stream->cfg.word_size / 8;
+	}
+
+	if ((mem_block_size <= DMA_MAX_TRANSFER_COUNT * data_size) && stream->reload_en) {
+		ret = dma_reload(dma_dev, stream->dma_channel, (uint32_t)stream->mem_block,
+				 (uint32_t)&cfg->reg->I2S_TXDMA, mem_block_size);
+	} else {
+		ret = i2s_siwx91x_prepare_dma_channel(i2s_dev, stream->mem_block, mem_block_size,
+						      stream->dma_channel, true);
+		stream->reload_en = false;
+	}
+	if (ret < 0) {
+		goto tx_disable;
+	}
+
+	ret = dma_start(dma_dev, stream->dma_channel);
+	if (ret < 0) {
+		goto tx_disable;
+	}
+
+	return;
+
+tx_disable:
+	i2s_siwx91x_stream_disable(stream, dma_dev);
+}
+
+static int i2s_siwx91x_param_config(const struct device *dev, enum i2s_dir dir)
+{
+	const struct i2s_siwx91x_config *cfg = dev->config;
+	struct i2s_siwx91x_data *data = dev->data;
+	struct i2s_siwx91x_stream *stream;
+	ULP_I2S_CLK_SELECT_T ref_clk;
+	uint32_t bit_freq;
+	uint32_t freq;
+	int resolution;
+	int ret;
+
+	if (dir == I2S_DIR_RX) {
+		stream = &data->rx;
+	} else {
+		stream = &data->tx;
+	}
+
+	resolution = i2s_siwx91x_convert_to_resolution(stream->cfg.word_size);
+	if (resolution < 0) {
+		return -EINVAL;
+	}
+
+	if (resolution != data->current_resolution) {
+		/* Configure primary mode and bit clock frequency */
+		bit_freq = 2 * stream->cfg.frame_clk_freq * stream->cfg.word_size;
+
+		if (cfg->reg == I2S0) {
```

Comment:
Let's try to not test the instance number.

BTW, I believe all the code related to the clock should goes in the clock driver. You can probably rely on `clock_control_set_rate()` to change the clock frequency dynamically.